### DISCUSSION TITLE
Understanding distant object movement when actuating joints

### DISCUSSION THREAD

**Discussion by hartikainen** [google-deepmind/mujoco#2098] [Asking for Help] [answered]
### Intro

Hi!

I am a MuJoCo user working on manipulation.

### My setup

MuJoCo:
```sh
$ python -c "import mujoco; print(mujoco.__version__)"
3.2.3
```

API:
C

OS:
```sh
python -c "import platform; print('OS:', platform.system(), platform.release(), platform.machine())"
OS: Darwin 24.0.0 arm64
```

### My question

I have a simple model (see below) which includes two entities: 1) a 3-link capsule-thingy with its "tail" position fixed to the worldbody, and 2) a free-moving cube. Things look as expected when I reset the model, but once I start adding actuator noise, the cube starts bouncing despite none of the actuated bodies touching the cube at all. I assume that this is not a bug, and I'm trying to understand the underlying reason for this effect. My question is: what is the mechanism through which the bodies interact with each other?

Here's a video of what's happening:

https://github.com/user-attachments/assets/ca66627d-0668-404f-9446-04248b4824e2

To reproduce this behavior, drag the xml-file from below into the MuJoCo simulate app and increase the noise scale.

Here are a couple of observations I've made when trying to understand this:
1. The Conjugate Gradient solver is much more susceptible to this behavior. The Newton solver reduces this effect significantly and, in this model, even seems to remove it altogether. However, I have another, a bit more complicated model (with effectively just the capsules substituted with Shadow Dex-EE-like fingers) where this effect occurs even with the Newton solver.
2. The solver parameters affect this behavior: e.g. with the more complicated Dex-EE model, going from `solref[1] == 0.9` to `solref[0] == 0.99` stabilizes the cube (at least visually).
3. The `iterations` and `ls_iterations` parameters also affect this. Increasing these parameters in the example model reduces the effect.
4. The mass of the geoms is inversely correlated with the magnitude effect. For example, setting `mass="0.01"` makes the simulation divergece whereas `mass="100"` reduces the jitter of the cube.

### Minimal model and/or code that explain my question

Model:
<details>
<summary>minimal XML</summary>

```XML
<mujoco model="scene">
  <option solver="CG" iterations="5" ls_iterations="5"/>
  <worldbody>
    <light pos="0 0 1.5" dir="0 0 -1"/>
    <geom name="floor" type="plane" size="0 0 0.05" />
    <body name="finger-base" pos="0 0 0.6" quat="0 1 0 0">
      <joint name="joint-0" type="hinge" axis="0 1 0" range="-0.785398 0.785398"/>
      <geom name="geom-0" type="capsule" contype="1" conaffinity="1" condim="6" size="0.0125" mass="1" rgba="1 0 0 1" fromto="0 0 0 0 0 0.1"/>
      <body name="body-0" pos="0 0 0.1">
        <joint name="joint-1" type="hinge" axis="0 1 0" range="-0.785398 0.785398"/>
        <geom name="geom-1" type="capsule" contype="1" conaffinity="1" condim="6" size="0.0125" mass="1" rgba="1 0 0 1" fromto="0 0 0 0 0 0.1"/>
        <body name="body-1" pos="0 0 0.1">
          <joint name="joint-2" type="hinge" axis="0 1 0" range="-0.785398 0.785398"/>
          <geom name="geom-2" type="capsule" contype="1" conaffinity="1" condim="6" size="0.0125" mass="1" rgba="1 0 0 1" fromto="0 0 0 0 0 0.1"/>
          <body name="body-2" pos="0 0 0.1"/>
        </body>
      </body>
    </body>
    <body name="cube" pos="0 0 0.025" quat="0 0 0 1">
      <geom name="cube" type="box" contype="1" conaffinity="1" condim="6" size="0.025 0.025 0.025" rgba="1 0 0 1" />
      <freejoint name="cube"/>
    </body>
  </worldbody>
  <actuator>
    <motor name="actuator-0" ctrllimited="true" ctrlrange="-1 1" gear="100" joint="joint-0"/>
    <motor name="actuator-1" ctrllimited="true" ctrlrange="-1 1" gear="100" joint="joint-1"/>
    <motor name="actuator-2" ctrllimited="true" ctrlrange="-1 1" gear="100" joint="joint-2"/>
  </actuator>
</mujoco>
```
</details>

### Confirmations

- [X] I searched the [latest documentation](https://mujoco.readthedocs.io/en/latest/overview.html) thoroughly before posting.
- [X] I searched previous [Issues](https://github.com/google-deepmind/mujoco/issues) and [Discussions](https://github.com/google-deepmind/mujoco/discussions), I am certain this has not been raised before.

**Comment by hartikainen** [comment]
Here's another, more surprising, observation: toggling the [`island`-flag](https://mujoco.readthedocs.io/en/stable/XMLreference.html#option-flag-island) affects this behavior, despite the documentation explicitly stating that the islands should have no effect on the physics:
> The flag currently has no effect on the physics pipeline, but enabling it allows for [island visualization](https://youtu.be/Vc1tq0fFvQA).

Here's a video of what happens:

https://github.com/user-attachments/assets/9c62987e-fb8a-45a6-aea8-e5623d601bd3

**Comment by yuvaltassa** [comment, marked as answer]
Answered in #2099

TL;DR, let the solver converge

**Reply by hartikainen** [reply]
Thanks, Yuval! This seems to be the critical part (from [#20](https://github.com/google-deepmind/mujoco/issues/2099#issuecomment-2379101082)):
> This is particularly egregious with CG, which—until the minimum is reached—will mix accelerations from unrelated dofs. 

Just to confirm: the object moving similarly at a distance when using the Newton solver is also expected, but would happen due to other inaccuracies in the solver rather than mixed accelerations between dofs?

**Reply by yuvaltassa** [reply]
Technically it could also happen with Newton solver if you don't let it converge, it's simply much better at converging...

Basically this could happen in any solver using the primal formulation. Only PGS is properly immune to this.
